[Tentative] Moving slow tokenizer to the Trie world.

Narsil opened the pull request:
# What does this PR do?

This PR attempts to solve the slow tokenizer `added_tokens` source of slowness.

Currently the splitting is done in O(n) manner, with very non obvious algorithm
to "pre-tokenize" (`tokenize` function). This will yield extremely slow tokenization
even by slow tokenization standards. It also affects slow-only tokenizers like ByT5.

The proposed fix simply moves the splitting into a O(1) algorithm (relative to `added_tokens`).
It does that by manually implementing a real Trie (more information why Python regexp can't be trusted on this: https://stackoverflow.com/questions/42742810/speed-up-millions-of-regex-replacements-in-python-3).

There is at least one know breaking change here, it's that users could rely on token ORDER to force splitting
on some `added_tokens` before others (https://github.com/huggingface/tokenizers/issues/615).
This won't be the case anymore with this code, as the splitting will happen on ~~first~~ longest encounter of `added_tokens` regardless. This is a pretty standard practice.

~~We could instead split on longest match first, but it's also a breaking change (although most likely less breaking).It does mean adding backtracking so the algorithm will be more complex and more state management~~ Edit: Implemented 


Benchmarking code:
```python
import datetime                                                                     
from transformers import GPT2Tokenizer                                                                     
                                                                                    
                                                                                                    
# They used to have to be sorted in reverse by length, otherwise the tokens arent           
newtokens = range(0, 20000)                                                          
newtokens = list(newtokens)                                                          
newtokens.sort(reverse=True)                                                         
newtokens = [f"new_{x}" for x in newtokens]                                       
                                                                                     
slow = GPT2Tokenizer.from_pretrained("gpt2")                                         
                                                                                     
# Add new vocab                                                                      
slow_custom = GPT2Tokenizer.from_pretrained("gpt2")                                 
slow_custom.add_tokens(newtokens)                                                               
                                                                                     
# Differences when tokenising the text...                                           
text = "this is a sentence containing new_200"                                     
for tokenizer in [slow, slow_custom]:                                                
    start = datetime.datetime.now()                                                  
    print(tokenizer.tokenize(text))                                                 
    print(datetime.datetime.now() - start) 
```

This goes from 4~7s on the `slow_custom` to 1ms (and ~0.3ms without `added_tokens`)




<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes https://github.com/huggingface/tokenizers/issues/615 (unrelated, because users seem to still
be using slow tokenizers there.


@LysandreJik
@patrickvonplaten
@n1t0

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->
```bash
RUN_SLOW=1 pytest -sv tests/test_tokenization_*

Results (1250.63s):
    3957 passed
     353 skipped
```

Narsil committed:
Moving slow tokenizer to the Trie world.

patrickvonplaten commented on src/transformers/tokenization_utils.py:
```diff
@@ -49,6 +49,84 @@
+    def __init__(self):
+        self.data = {}
+
+    def add(self, word):
```
(nit) a bit more docstring of what is happening here would be nice

patrickvonplaten commented on src/transformers/tokenization_utils.py:
```diff
@@ -49,6 +49,84 @@
+            ref = ref[char]
+        ref[""] = 1
+
+    def split(self, text):
```
(nit) Also pretty hard to understand what's happening here IMO - maybe add some more docstring as well?

patrickvonplaten reviewed (commented):
In general I'm in favor of this I think! I haven't looked too much into the code, but:

- if the breaking changes are kept to only exotic cases like you mentioned
- more tests are added (especially for the `Trie.split(...)` function) => this also helps a lot to understand the code better

Also, I haven't fully understood if the PR speeds up tokenization only if *a lot* of tokens are added via `add_tokens(...)` or does it speed up tokenization in general? IMO adding 2000 tokens via `add_tokens(...)` is a rather exceptional case no?

Narsil commented:
@patrickvonplaten It speeds up all the time, but it's only that impressive with adding tons of `added_tokens`. The reason is because of the change in complexity. Even for simple `ByT5` that uses 125 extra `added_tokens` by default, that's already 100x speed (so not really that exotic). 

For "regular" slow tokenizers, with <5 added_tokens the speedup exists but is rather negligible.

Also I've seen 2/3 different issues regarding this. The one reported (https://github.com/huggingface/tokenizers/issues/615) has 8 participants since February, so while not super urgent, there definitely seems to be more than a couple of people doing that.

And always fine adding more documentation, but fwiw it's a pretty standard data structure.

Narsil committed:
Adding more docstrings to the Trie.

Narsil committed:
Fixing doctest (incompatible wiht our format? )

LysandreJik commented on src/transformers/tokenization_utils.py:
```diff
@@ -135,6 +255,7 @@ def __init__(self, **kwargs):
         self.added_tokens_encoder: Dict[str, int] = {}
         self.added_tokens_decoder: Dict[int, str] = {}
         self.unique_no_split_tokens: List[str] = []
+        self.trie = Trie()
```
Would it be possible to have a more explicit name of the property? For example `unique_tokens_trie`?

Narsil replied:
I don't like `unique` in there, because there's nothing related to uniqueness in there.

`no_split_trie` could work, but we're actually using it to split around some tokens.
`tokens_trie` could work IMO.
`no_split_tokens_trie` is a bit longer and not really informative (there are no `split_tokens` afaik).

LysandreJik reviewed (commented):
Thank you for implementing this! Would it be possible to add some tests so that we could identify:
1. What this enables vs the previous implementation
2. The changes in behavior this will bring, i.e tests that would break with the current implementation that the proposed implementation will solve

Thank you!

Narsil commented:
Hi @LysandreJik Do we have benchmarks  tests anywhere ?

This doesn't fix anything that was broken before, it just make things faster (some usage of the lib was so slow that it was unusable, but it was definitely working).

I could add a test that some tokenization takes too long, but it's always a tricky business to add tests related to timings because it might depend on the hardware running the tests, so it would definitely NOT be a unit test.

sgugger commented on src/transformers/tokenization_utils.py:
```diff
@@ -279,87 +410,40 @@ def tokenize(self, text: TextInput, **kwargs) -> List[str]:
-                # Cf. https://github.com/huggingface/transformers/pull/2778
-                # and https://github.com/huggingface/transformers/issues/3788
+        no_split_token = set(self.unique_no_split_tokens)
+        trie = self.trie
+        tokens = trie.split(text)
```
Why not directly?
```suggestion
        tokens = self.trie.split(text)
```
It doesn't look like `trie` is used anywhere else.

Narsil replied:
Not sure either why I wrote that in 2 steps :)

sgugger commented on src/transformers/tokenization_utils.py:
```diff
@@ -49,6 +49,126 @@
+        offsets = []
+
+        skip = None
+        for current, c in enumerate(text):
+            if skip and current < skip:
+                # Prevents the lookahead for matching twice
+                # like extra_id_100 and id_100
+                continue
+            to_remove = set()
+            reset = False
+            for start, p in states.items():
+                if c in p:
+                    p = p[c]
+                    if "" in p:
+                        # Lookahead to match longest first
+                        # Important in case of extra_id_1 vs extra_id_100
+                        lookahead_index = current + 1
+                        end = current + 1
+                        next_c = c
+                        while next_c in p:
+                            p = p[next_c]
+                            lookahead_index += 1
+                            if "" in p:
+                                end = lookahead_index + 1
+                                skip = lookahead_index + 1
+
+                            if lookahead_index == len(text):
+                                # End of string
+                                break
+                            next_c = text[lookahead_index]
+                        # End lookahead
+
+                        offsets.append(start)
+                        offsets.append(end)
+                        reset = True
+                    states[start] = p
+                else:
+                    to_remove.add(start)
```
Is there any way to replace the variable name by something more explicit? This is super hard to read and understand.
More comments would also be welcome.

Narsil replied:
I can change `c` into `current_char` and `p` into `trie_pointer` maybe ?

Overall state machines code is always a bit hard to read (and complexified by the lookahead part).
I don't know a great way to make this a usual python-readable nice.

I think the best way to improve readability is to document a bit more what each variable should be, or do you think variable names are the bigger culprits ?

sgugger replied:
At least knowing c is a char helps tremendously, so all for the variable renames! And a few comments on what is happening would be amazing as well if you can.

Narsil replied:
I added lots of comments everywhere, hopefully explaining a bit more how every piece is supposed to work and interact with each other.

sgugger replied:
Thanks, it's very helpful!

sgugger commented on src/transformers/tokenization_utils.py:
```diff
@@ -49,6 +49,126 @@
+
+        This trie will match the longest possible word first !
+
+        >>> trie = Trie()
+
+        >>> trie.split("[CLS] This is a extra_id_100")
+
+        ["[CLS] This is a extra_id_100"]
+
+        >>> trie.add("[CLS]")
+
+        >>> trie.add("extra_id_1")
+
+        >>> trie.add("extra_id_100")
+
+        >>> trie.split("[CLS] This is a extra_id_100")
+
+        ["[CLS]", " This is a ", "extra_id_100"]
```
No need for all the whitespace.
```suggestion
        >>> trie = Trie()
        >>> trie.split("[CLS] This is a extra_id_100")
        ["[CLS] This is a extra_id_100"]

        >>> trie.add("[CLS]")
        >>> trie.add("extra_id_1")
        >>> trie.add("extra_id_100")
        >>> trie.split("[CLS] This is a extra_id_100")
        ["[CLS]", " This is a ", "extra_id_100"]
```

Narsil replied:
Actually this breaks the doc formatting tool. And fails the quality test. But indeed it would be better.

sgugger replied:
I think this is because you did not put it in an `Example::` block.

Narsil replied:
Make sense, done.

sgugger commented on src/transformers/tokenization_utils.py:
```diff
@@ -49,6 +49,126 @@
+        Passes over every char (utf-8 char) on word and recursively adds it to the internal `data` trie representation.
+        The special key `""` is used to represent termination.
+
+        This function is idempotent, adding twice the same word will leave the trie unchanged
+
+        >>> trie = Trie()
+
+        >>> trie.add("Hello 友達")
+
+        >>> trie.data
+
+        {"H": {"e": {"l": {"l": {"o": {" ": {"友": {"達": {"": 1}}}}}}}}}
+
+        >>> trie.add("Hello")
+
+        >>> trie.data
+
+        {"H": {"e": {"l": {"l": {"o": {"": 1, " ": {"友": {"達": {"": 1}}}}}}}}}
```
```suggestion
        This function is idempotent, adding twice the same word will leave the trie unchanged.

        >>> trie = Trie()
        >>> trie.add("Hello 友達")
        >>> trie.data
        {"H": {"e": {"l": {"l": {"o": {" ": {"友": {"達": {"": 1}}}}}}}}}

        >>> trie.add("Hello")
        >>> trie.data
        {"H": {"e": {"l": {"l": {"o": {"": 1, " ": {"友": {"達": {"": 1}}}}}}}}}
```
No need for all those new lines.

sgugger commented on src/transformers/tokenization_utils.py:
```diff
@@ -49,6 +49,126 @@
 
+class Trie:
+    """
+    Trie in Python. Creates a Trie out of a list of words. The trie is used to split on `added_tokens` in one pass
```
Link to the algorithm for those who don't know it? -> https://en.wikipedia.org/wiki/Trie

sgugger reviewed (commented):
I understand the generic concept and the speedup is definitely welcome! However I can't understand the internal implementation as it is, which is going to make it hard to maintain.

Narsil committed:
Update src/transformers/tokenization_utils.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Narsil committed:
Adding a lot more comment into the internals of this algorithm.

Narsil committed:
Cleaner doc.

Narsil committed:
Fixing the namings.

LysandreJik commented on src/transformers/tokenization_utils.py:
```diff
@@ -49,6 +49,172 @@
+
+        This trie will match the longest possible word first !
+
+        Example::
```
To have this correctly rendered in the docs, an additional line return is necessary
```suggestion
        Example::
        
```

LysandreJik reviewed (commented):
I understand what you mean regarding tests - one that comes to mind is the breaking change you mention:

> This won't be the case anymore with this code, as the splitting will happen on first longest encounter of added_tokens regardless. This is a pretty standard practice.

A test ensuring this is the case now would be nice. This identifies this as a property of our tokenizers that is known and supported, rather than a side-effect.

Narsil committed:
Update src/transformers/tokenization_utils.py

Co-authored-by: Lysandre Debut <lysandre@huggingface.co>

Narsil committed:
quality.

Narsil committed:
Fixing longest first match.

Narsil committed:
Small improvements to cuts + more test + canine resistant test.

Narsil committed:
Fixing fast test.

Narsil commented:
@LysandreJik Ok, I added 1 `common` test which fails only on Canine (`extra_id_1` is not valid over there).

Also added Trie specific tests (the one in the doc basically)

SaulLu reviewed (commented):
Thank you very much for this work and this addition! It's great to have found a way to speed up the tokenization of slow tokenizers.

After all the comments that have been made, I don't have much to add. Except that after renaming the variables and adding the comments I find the method very readable and understandable! Thanks also for the information about the unreliability of `Re` for this type of method.

Regarding the breaking changes, I would just like to make sure that everyone has in mind the current behaviour and the behaviour induced by this PR. As I find examples quite informative, I'm sharing two notebooks both comparing the behaviour of the `add_tokens` method for slow and fast tokenizers: but one uses the [latest release of transformers](https://colab.research.google.com/drive/1scfWUO0GV7i1zaxvescSOD8TUSYRk6IC?usp=sharing) and the other [@Narsil branch](https://colab.research.google.com/drive/1mtLvjaU9fBhFq6VKsHd64p4GDBfNSO1_?usp=sharing).  

Initially I didn't immediately see the complexity of this task (see [this comment](https://github.com/huggingface/tokenizers/issues/772)) but in the end my opinion is that fast tokenizers offer the best API for adding tokens. Currently, fast and slow tokenizers do not behave the same and it seems to me that after this PR they will still not behave the same (but not in the same way). My concern is that the difference that would now be between slow and fast tokenizers does not seem simple to understand for a new user. What do you think?

Narsil commented:
Thanks @SaulLu , the matching was incorrect in that edge case where some token is rigourously included in another, then we would match the inner token instead of the first match.

The lookahead part got more complex but will now work in that edge case (which is important to at least follow the documentation)

Regarding the idea of `added_tokens` following order, recaping some offline conversation:
- It's doable, but would make code even more complex. We would need to keep track of ranks in the Trie, whenever we have a full match, resolve all partial matches, sort by order and take the highest rank.
- Seems overly complex for what seems to be pathological cases at best, so out of scope of this one.

Narsil commented:
Will merge this later today unless there are still some comments (But I feel it's ok in current state)
